### PR TITLE
CAMEL-15496: Add support for additional headers and properties

### PR DESCRIPTION
**Pull request by rgannu**
The additional headers will be added on top of the existing headers
from the message.

The properties added are only basic AMQP properties as defined in the
 {@link com.rabbitmq.client.AMQP.BasicProperties.Builder}. When the
message contains already these properties then will be considered and
these additional properties will be ignored.

Introduced a SimpleDataHolderBean to hold a hashmap for adding any
header. Changed the RabbitMQConstants to Enum so that we can also add
the property description in that.

The rabbitmq AMQP basic properties constant keys should not
have dot in their prefixes. This gives problem in trying to
retrieve the value due to fix of CAMEL-15394.

As with this fix though we can add any header, only static/hard-coded
values could be associated.

### CODE REVIEW

File: archetypes/camel-archetype-api-component/src/main/resources/archetype-resources/__artifactId__-api/pom.xml [hunk 1]
```diff
@@ -78,7 +78,7 @@
             <groupId>org.apache.maven.plugins</groupId>
             <artifactId>maven-javadoc-plugin</artifactId>
             <configuration>
-              <additionalparam>-Xdoclint:none</additionalparam>
+              <doclint>none</doclint>
```
Comment (oscerd) [inline archetypes/camel-archetype-api-component/src/main/resources/archetype-resources/__artifactId__-api/pom.xml:81]: This seems to be wrong

File: components/camel-as2/camel-as2-api/pom.xml [hunk 2]
```diff
@@ -127,7 +127,7 @@
                         <groupId>org.apache.maven.plugins</groupId>
                         <artifactId>maven-javadoc-plugin</artifactId>
                         <configuration>
-                            <additionalparam>-Xdoclint:none</additionalparam>
+                            <doclint>none</doclint>
```
Comment (oscerd) [inline components/camel-as2/camel-as2-api/pom.xml]: Seems to be not related to this PR, please rebuild the whole project

File: components/camel-box/camel-box-api/pom.xml [hunk 3]
```diff
@@ -87,7 +87,7 @@
                         <groupId>org.apache.maven.plugins</groupId>
                         <artifactId>maven-javadoc-plugin</artifactId>
                         <configuration>
-                            <additionalparam>-Xdoclint:none</additionalparam>
+                            <doclint>none</doclint>
```
Comment (oscerd) [inline components/camel-box/camel-box-api/pom.xml]: Ditto

File: components/camel-fhir/camel-fhir-api/pom.xml [hunk 4]
```diff
@@ -92,7 +92,7 @@
                         <groupId>org.apache.maven.plugins</groupId>
                         <artifactId>maven-javadoc-plugin</artifactId>
                         <configuration>
-                            <additionalparam>-Xdoclint:none</additionalparam>
+                            <doclint>none</doclint>
```
Comment (oscerd) [inline components/camel-fhir/camel-fhir-api/pom.xml]: Ditto

File: components/camel-olingo2/camel-olingo2-api/pom.xml [hunk 5]
```diff
@@ -122,7 +122,7 @@
                         <groupId>org.apache.maven.plugins</groupId>
                         <artifactId>maven-javadoc-plugin</artifactId>
                         <configuration>
-                            <additionalparam>-Xdoclint:none</additionalparam>
+                            <doclint>none</doclint>
```
Comment (oscerd) [inline components/camel-olingo2/camel-olingo2-api/pom.xml:125]: This seems wrong, please rebuild the whole project

File: components/camel-olingo4/camel-olingo4-api/pom.xml [hunk 6]
```diff
@@ -136,7 +136,7 @@
                         <groupId>org.apache.maven.plugins</groupId>
                         <artifactId>maven-javadoc-plugin</artifactId>
                         <configuration>
-                            <additionalparam>-Xdoclint:none</additionalparam>
+                            <doclint>none</doclint>
```
Comment (oscerd) [inline components/camel-olingo4/camel-olingo4-api/pom.xml:139]: This seems wrong, please rebuild the whole project

File: components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQConstants.java [hunk 7]
```diff
@@ -16,42 +16,76 @@
...
-    public static final String RABBITMQ_DIRECT_REPLY_ROUTING_KEY = "amq.rabbitmq.reply-to";
-    public static final String RABBITMQ_QUEUE_LENGTH_LIMIT_KEY = "x-max-length";
-    public static final String RABBITMQ_QUEUE_MAX_PRIORITY_KEY = "x-max-priority";
-    public static final String RABBITMQ_QUEUE_MESSAGE_TTL_KEY = "x-message-ttl";
-    public static final String RABBITMQ_QUEUE_TTL_KEY = "x-expires";
-    public static final String RABBITMQ_QUEUE_SINGLE_ACTIVE_CONSUMER_KEY = "x-single-active-consumer";
-
-    private RabbitMQConstants() {
-        // Constants class
+    ROUTING_KEY("rabbitmq-ROUTING_KEY", "The routing key that will be used when sending the message"),
```
Comment (oscerd) [inline components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQConstants.java:26]: This will breaks existing users and I don't see a good reason for this

Comment (oscerd) [reply components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQConstants.java:26]: Also you're changing the headers name.

**Review by oscerd — CHANGES_REQUESTED** [review]
I think there are updates not needed for the purpose of this PR. @davsclaus please can you have a look? Since you reviewed the one in 3.4.x (closed https://github.com/apache/camel/pull/4155/files)

**Comment by rgannu** [comment]
I am using IntelliJ and I couldn't format the files.. 
I imported the code-style from `buildingtools/src/main/resources/camel-eclipse-formatter-config.xml`

`mvn install -Psourcecheck` always fails. 

> [ERROR] Failed to execute goal org.apache.camel:camel-format-plugin:3.6.0-SNAPSHOT:validate (validate) on project camel-core: File '/home/ganesh/IdeaProjects/github/camel-unifly/core/camel-core/src/test/java/org/apache/camel/util/EndpointHelperTest.java' has not been previously formatted.  Please format file and commit before running validation! -> [Help 1]

Can you please let me know which code-formatter is used by all ?

File: components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQConstants.java (hunk 7 above)
Comment (rgannu) [reply components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQConstants.java:26]: Header names are not changed.. I checked myself... 
For my PR, I added support to add via the sink connector configuration way to add any Basic AMQP properties as 

 `    
"camel.sink.endpoint.additionalProperties": "#bean:addProperties",

"camel.beans.addProperties": "#class:org.apache.camel.support.beans.SimpleDataHolderBean",
"camel.beans.addProperties.mapData[rabbitmq-CONTENT_TYPE]": "application/octet-stream",
"camel.beans.addProperties.mapData[rabbitmq-CONTENT_ENCODING]": "binary",
`
When the property value is given as `mapData[rabbitmq.CONTENT_TYPE]`, due to the fix done as part of CAMEL-15394, 
https://github.com/apache/camel/blob/master/core/camel-support/src/main/java/org/apache/camel/support/PropertyBindingSupport.java#L408
the property is split into parts and then we are trying to search for property value for `CONTENT_TYPE]` which will not be there. 

I do not want to change this part and so I can in RabbitMQConstants.java
 - Remove the prefix `rabbitmq.` or
 - Change the prefix to `rabbitmq-` (which is what I have done)
Both of these solution works, without changing the changes done for CAMEL-15394.

Also the properties get set properly in the AMQP message as this is based on the com.rabbitmq.client.AMQP.BasicProperties in which we need to explicitly set the basic AMQP properties.

Comment (oscerd) [reply components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQConstants.java:26]: Rabbitmq. Was the original prefix, now it's rabbitmq-.. anyone using the header explicitly will have his route broken.

Comment (oscerd) [reply components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQConstants.java:26]: You don't have to think only about camel kafka connector, this is something related to pure camel too

Comment (oscerd) [reply components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQConstants.java:26]: so it's -1 for me.

**Comment by oscerd** [comment]
> I am using IntelliJ and I couldn't format the files..
> I imported the code-style from `buildingtools/src/main/resources/camel-eclipse-formatter-config.xml`
> 
> `mvn install -Psourcecheck` always fails.
> 
> > [ERROR] Failed to execute goal org.apache.camel:camel-format-plugin:3.6.0-SNAPSHOT:validate (validate) on project camel-core: File '/home/ganesh/IdeaProjects/github/camel-unifly/core/camel-core/src/test/java/org/apache/camel/util/EndpointHelperTest.java' has not been previously formatted.  Please format file and commit before running validation! -> [Help 1]
> 
> Can you please let me know which code-formatter is used by all ?

you just need to build the entire project and it will be auto-formatted.

File: components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQConstants.java (hunk 7 above)
Comment (rgannu) [reply components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQConstants.java:26]: > Rabbitmq. Was the original prefix, now it's rabbitmq-.. anyone using the header explicitly will have his route broken.
I understand completely. 
Earlier my PR to support adding property with `mapData[rabbitmq.CONTENT_TYPE]` was working fine in camel-3.4.0.

I had to submit the PR against the master and this is an enhancement work. 
When I did the changes against the master, I started getting problems that the none of the properties getting their value set due to the changes done in CAMEL-15394. Now, please let me know how I can fix this. Without the changes to the prefix so that it doesn't contain the dot (.) character, I couldn't submit this PR.

Comment (oscerd) [reply components/camel-rabbitmq/src/main/java/org/apache/camel/component/rabbitmq/RabbitMQConstants.java:26]: now I'm getting the point.

File: components/camel-as2/camel-as2-api/pom.xml (hunk 2 above)
Comment (rgannu) [reply components/camel-as2/camel-as2-api/pom.xml]: Agreed. Thanks. 
I will remove these changes.

File: components/camel-box/camel-box-api/pom.xml (hunk 3 above)
Comment (rgannu) [reply components/camel-box/camel-box-api/pom.xml]: Agreed. Thanks. 
I will remove these changes.

File: components/camel-olingo2/camel-olingo2-api/pom.xml (hunk 5 above)
Comment (rgannu) [reply components/camel-olingo2/camel-olingo2-api/pom.xml]: Agreed. Thanks. 
I will remove these changes.

File: components/camel-olingo4/camel-olingo4-api/pom.xml (hunk 6 above)
Comment (rgannu) [reply components/camel-olingo4/camel-olingo4-api/pom.xml]: Agreed. Thanks. 
I will remove these changes.

**Comment by rgannu** [comment]
Even after repeated attempts to rebase and merge I am getting failures in the checkstyle step. 
I will close this PR and open a new PR from master.
